python/lsst/ts/FiberSpectrograph/dataManager.py

```diff
@@ -0,0 +1,130 @@
+    temperature_setpoint: float
+    """The internal spectrograph temperature set point."""
+    n_pixels: int
+    """The number of pixels in the detector."""
+
+
+class DataManager:
+    """A data packager and uploader for output from the Fiber Spectrograph CSC.
+
+    Attributes
+    ----------
+    telescope : `str`
+        The name of the telescope the instrument is attached to.
+    instrument : `str`
+        The name of the instrument taking the data.
+    outpath : `str`, optional
+        A path to write the FITS files to.
+        **WARNING**: this argument will be removed once the API for writing
+        to the LFA becomes available in salobj; do not rely on it.
+    """
+    def __init__(self, telescope, instrument, outpath=None):
+        self.telescope = telescope
+        self.instrument = instrument
+        self.outpath = outpath
+
+    def __call__(self, data):
+        """Generate a FITS hdulist built from SpectrographData.
+
+        Parameters
+        ----------
+        data : `SpectrographData`
+            The data to build the FITS output with.
+
+        Returns
+        -------
+        uri : `str`
+            The file path or other identifier (e.g. s3 store) for where the
+            data was sent.
+        """
+        hdu1 = self.make_primary_hdu(data)
+        hdu2 = self.make_wavelength_hdu(data)
+        hdulist = astropy.io.fits.HDUList([hdu1, hdu2])
+        # TODO: this writeto(file) block should be removed once we have a
+        # working LFA API to send data to.
+        if self.outpath is not None:
+            filename = os.path.join(self.outpath, f"{self.instrument}_{data.date_begin.tai.fits}.fits")
+            hdulist.writeto(filename, checksum=True)
+            return filename
+        else:
+            return None
+
+    def make_fits_header(self, data):
+        """Return a FITS header built from SpectrographData."""
+        hdr = astropy.io.fits.Header()
+        # TODO: it would be good to include the dataclass docstrings
+        # as comments on each of these...
+        hdr['TELESCOP'] = self.telescope
+        hdr['INSTRUME'] = self.instrument
+        hdr['DETSIZE'] = data.n_pixels
+        hdr['DATE-BEG'] = astropy.time.Time(data.date_begin).tai.fits
+        hdr['DATE-END'] = astropy.time.Time(data.date_end).tai.fits
+        hdr['EXPTIME'] = data.duration
+        hdr['TIMESYS'] = 'TAI'
+        hdr['IMGTYPE'] = data.type
+        hdr['SOURCE'] = data.source
+        hdr['TEMP_SET'] = data.temperature_setpoint
+        hdr['CCDTEMP'] = data.temperature
+        return hdr
+
+    def make_primary_hdu(self, data):
+        """Return the primary HDU build from SpectrographData."""
+        hdu = astropy.io.fits.PrimaryHDU(data=data.spectrum, header=self.make_fits_header(data))
+        return hdu
+
+    def make_wavelength_hdu(self, data):
+        """Return the wavelength HDU build from SpectrographData."""
+        hdu = astropy.io.fits.ImageHDU(data=data.wavelength)
```

Comment:
Is there any way we can put the units of this wavelength in metadata somewhere? If this is the wavelength of every pixel in the primary HDU then it could/should be stored as a standard WCS associated with the data itself (using the -TAB format if needs be; Greisen et al 2006 A&A, 446, 747).

Reply:
After having spent a few days digging into this, it looks like neither astropy nor specutils currently supports the spectral `-TAB` WCS format. You can make a `gwcs` tablular spectral wcs, but it doesn't support writing it to FITS as a G. 2006 style FITS WCS.

I think I might file another ticket and defer making a FITS WCS here until one of the above properly supports it.

Reply:
See this PR I made to specutils: https://github.com/astropy/specutils/pull/530
and the latest discussion on this astropy Issue: https://github.com/astropy/astropy/issues/7883

Reply:
I've changed to outputting the wavelength data as a BinTable HDU with unit information.

Given the fact that astropy and specutils don't fully support this type of WCS, I don't have a good way to verify that I've done it correctly if I try to do it with astshim. I can make the AST FrameDict, but I don't know how to turn that into a Greisen 2006-style WCS:

```python
pixelFrame = astshim.Frame(1, "Domain=PIXELS");
wavelengthFrame = astshim.Frame(1, "Domain=WAVELENGTH");
wavelength = np.arange(300, 1000, 1000)
lutMap = astshim.LutMap(wave, 0, 1)
frameDict = astshim.FrameDict(pixelFrame)
frameDict.addFrame("PIXELS", lutMap, wavelengthFrame)
```

Reply:
Look up the TabOK parameter in AST for when you want to generate the FitsChan from the WCS.